<!DOCTYPE html>
<html>
  <head>
    <title>3D Genome Data Procesing, Analysis, and Visualization</title>
    <meta charset="utf-8">
    <style>
      @import url(https://fonts.googleapis.com/css?family=Yanone+Kaffeesatz);
      @import url(https://fonts.googleapis.com/css?family=Droid+Serif:400,700,400italic);
      @import url(https://fonts.googleapis.com/css?family=Ubuntu+Mono:400,700,400italic);

      body { font-family: 'Droid Serif'; }
      h1, h2, h3 {
        font-family: 'Yanone Kaffeesatz';
        font-weight: normal;
      }
      .remark-code, .remark-inline-code { font-family: 'Ubuntu Mono'; }
    </style>
  </head>
  <body>
    <textarea id="source">

class: center, middle

# 3D Genome Data Processing, Analysis, and Visualization Tutorial

## ISMB 2017, Prague, Czech Republic

#### Nezar Abdennur, Soo Lee, Nils Gehlenborg, Peter Kerpedjiev, Jian Ma

---

# Agenda

1. Introduction
2. Deep-dive
3. ...

---

# Example pipeline

Starting from SRR1658602_{1,2}.fastq

---

# Practice!
Hi-C data processing pipeline

1. Mapping
  * Tools : `bwa`, `samtools`
  * output: `bam` file

2. Filtering / sorting / Creating a list of contacts
  * Tools: `pairsamtools`, `pairix`
  * outputs: `pairs`, `bam` files

3. Binning
  * Tools: `cooler`
  * outputs: `cool` file

4. Normalization
  * Tools: `cooler`
  * outputs: `cool` file

---

# Alignment!

  * Tools : 
    * `bwa` (http://bio-bwa.sourceforge.net/) : alignment
    * `samtools` (https://samtools.github.io/) : format conversion (sam->bam)

  * Command
    ```
    bwa mem -SP5M <index> <fastq1> <fastq2> | samtools view -bhS - > output.bam
    ```

  * input: `fastq` files, bwa genome index files

  * output: `bam` file

---

# How to read a `bam` file

  * A `bam` file is a binary file that stores alignment information for sequenced reads

  * Tools : 
    * `samtools` (https://samtools.github.io/) : viewing a bam file

  * Command for viewing
    ```
    samtools view <bamfile> | less
    ```

  * SAM/BAM format: https://samtools.github.io/hts-specs/SAMv1.pdf

  * Example lines
  ```
  SRR1658581.203938	81	chr5	76758064	0	51M50S	=	76757789	-326	GGGGTTTTACCTTGTTGGCCAGGCTGGTCTTGAACTCCTGACCTCATGATCGATCATCACCATCTTGGTTTTGGTGGGTTTTGGCCGGCTTCCTTACTGCA	.BB@DCCC@DCDDBBBDDDDCDDDDDDDDEDCC>DECCEEEFFHGHHHJIGIJJJJHHFIIIJIIJHIJJJJJJJJJJJJJJJIJJJJHHHHHFFFFFCCC	NM:i:0	MD:Z:51	AS:i:51	XS:i:48	SA:Z:chr5,76780991,+,51M50S,0,0;
  SRR1658581.203938	2113	chr5	76780991	0	51M50H	=	76757789	-23203	TGCAGTAAGGAAGCCGGCCAAAACCCACCAAAACCAAGATGGTGATGATCG	CCCFFFFFHHHHHJJJJIJJJJJJJJJJJJJJJIHJIIJIIIFHHJJJJIG	NM:i:0	MD:Z:51	AS:i:51	XS:i:43	SA:Z:chr5,76758064,-,51M50S,0,0;
  SRR1658581.203938	161	chr5	76757789	60	101M	=	76758064	326	TTGATTTTTCCTTCTGGACAGGCTCCAATATGGCTCTGTGCAGCACTGTTATTGGTCCTGTCGTTACTGAAAATTCTGATTTTTTTTTTTTGGAAACGGAG	CCCFFFFFHHHHHJJJJJJJJJJEHJJJJIIIJFEIJIHIIJIIIGIIHHEIIEGHGHGIFHIHIGIGIIJJHGHHHH;?CEEDDDDDDDD<-4>AC?B85	NM:i:0	MD:Z:101	AS:i:101	XS:i:0

  ```

---


# Filtering!

  * Tools : 
    * `pairsamtools` (https://github.com/mirnylab/pairsamtools) : filtering

  * Command
    ```
    pairsamtools parse -c <chromsize_file> <bam_file> |
    { pairsamtools sort } | 
    { pairsamtools select '(pair_type == "CX") or (pair_type == "LL")' } |
    { pairsamtools dedup \
        --output \
            >( pairsamtools split \
                --output-pairs ${NODUPS_PAIRS_PATH} \
                --output-sam ${NODUPS_SAM_PATH} ) 
    }
    ```

  * input: `bam` file

  * output: `bam`, `pairs` file


---


# Pairs file
 
  * A `pairs` file is a block-gzipped text file that stores list of contact loci

  * Tools:
    * `pairix` (https://github.com/4dn-dcic/pairix) : indexing & querying `pairs` files

  * Command
    ```
    pairix <pairsfile> <querystr> [ <querystr> ... ] 
    ```
    ```
    pairix <pairsfile> 'chr1|chr1' 'chr2:1-2000000|chr4:5000000-6000000'
    ```

  * Pairs format: https://github.com/4dn-dcic/pairix/blob/master/pairs_format_specification.md

  * Example lines
  ```
  #columns: readID chr1 pos1 chr2 pos2 strand1 strand2
  EAS139:136:FC706VJ:2:2104:23462:197393 chr1 10000 chr1 20000 + +
  EAS139:136:FC706VJ:2:8762:23765:128766 chr1 50000 chr1 70000 + +
  EAS139:136:FC706VJ:2:2342:15343:9863 chr1 60000 chr2 10000 + +
  ```

---

# Binning!

  * Tools:
    * `cooler` (https://github.com/mirnylab/cooler)

  * Command
    ```
    cooler cload pairix <chrsize_file>:<bin_size> <pairs_file> <out_prefix>.cool
    ```

  * input: `pairs` file, `chromsize` file

  * output : `cool` file


---

# How to read a `cool` file

  * A `cool` file is an HDF5-format file that stores contact matrices and normailzation vectors.

  * Tools:
    * `cooler` (https://github.com/mirnylab/cooler)

  * Command
    ```
    cooler dump -b -t pixels --header --join -r chr3:1-12,000,000 -r2 chr17 out.cool 
    ```

  * Cool format : http://cooler.readthedocs.io/en/latest/datamodel.html


---

# Normalization! (Matrix balancing)

  * Tools:
    * `cooler` (https://github.com/mirnylab/cooler)

  * Command
    ```
    cooler balance <cool_file>
    ```

  * input : `cool` file
  
  * output : `cool` file


---


    </textarea>
    <script src="https://remarkjs.com/downloads/remark-latest.min.js">
    </script>
    <script>
      var slideshow = remark.create();
    </script>
  </body>
</html>
