<!DOCTYPE html>
<html>
  <head>
    <title>3D Genome Data Procesing, Analysis, and Visualization</title>
    <meta charset="utf-8">
    <style>
      @import url(https://fonts.googleapis.com/css?family=Yanone+Kaffeesatz);
      @import url(https://fonts.googleapis.com/css?family=Droid+Serif:400,700,400italic);
      @import url(https://fonts.googleapis.com/css?family=Ubuntu+Mono:400,700,400italic);

      body { font-family: 'Droid Serif'; }
      h1, h2, h3 {
        font-family: 'Yanone Kaffeesatz';
        font-weight: normal;
      }
      .remark-code, .remark-inline-code { font-family: 'Ubuntu Mono'; }
    </style>
  </head>
  <body>
    <textarea id="source">

class: center, middle

# 3D Genome Data Processing, Analysis, and Visualization Tutorial

## ISMB 2017, Prague, Czech Republic

#### Nezar Abdennur, Soo Lee, Nils Gehlenborg, Peter Kerpedjiev, Jian Ma

---

# Agenda

1. Introduction
2. Deep-dive
3. ...

---

# Practice!
Hi-C data processing pipeline

1. Mapping
  * Tools : `bwa`, `samtools`
  * output: `bam` file

2. Filtering / sorting / Creating a list of contacts
  * Tools: `pairsamtools`, `pairix`
  * outputs: `pairs`, `bam` files

3. Binning
  * Tools: `cooler`
  * outputs: `cool` file

4. Normalization
  * Tools: `cooler`
  * outputs: `cool` file

---

# Alignment!

  * Tools : 
    * `bwa` (http://bio-bwa.sourceforge.net/) : alignment
    * `samtools` (https://samtools.github.io/) : format conversion (sam->bam)

  * Command
    ```
    bwa mem -SP5M <index> <fastq1> <fastq2> | samtools view -bhS - > output.bam
    ```

  * input: `fastq` files, bwa genome index files

  * output: `bam` file

---

# How to read a `bam` file

  * A `bam` file is a binary file that stores alignment information for sequenced reads

  * Tools : 
    * `samtools` (https://samtools.github.io/) : viewing a bam file

  * Command for viewing
    ```
    samtools view <bamfile> | less
    ```

  * SAM/BAM format: https://samtools.github.io/hts-specs/SAMv1.pdf

  * Example lines
  ```
  SRR1658581.203938	81	chr5	76758064	0	51M50S	=	76757789	-326	GGGGTTTTACCTTGTTGGCCAGGCTGGTCTTGAACTCCTGACCTCATGATCGATCATCACCATCTTGGTTTTGGTGGGTTTTGGCCGGCTTCCTTACTGCA	.BB@DCCC@DCDDBBBDDDDCDDDDDDDDEDCC>DECCEEEFFHGHHHJIGIJJJJHHFIIIJIIJHIJJJJJJJJJJJJJJJIJJJJHHHHHFFFFFCCC	NM:i:0	MD:Z:51	AS:i:51	XS:i:48	SA:Z:chr5,76780991,+,51M50S,0,0;
  SRR1658581.203938	2113	chr5	76780991	0	51M50H	=	76757789	-23203	TGCAGTAAGGAAGCCGGCCAAAACCCACCAAAACCAAGATGGTGATGATCG	CCCFFFFFHHHHHJJJJIJJJJJJJJJJJJJJJIHJIIJIIIFHHJJJJIG	NM:i:0	MD:Z:51	AS:i:51	XS:i:43	SA:Z:chr5,76758064,-,51M50S,0,0;
  SRR1658581.203938	161	chr5	76757789	60	101M	=	76758064	326	TTGATTTTTCCTTCTGGACAGGCTCCAATATGGCTCTGTGCAGCACTGTTATTGGTCCTGTCGTTACTGAAAATTCTGATTTTTTTTTTTTGGAAACGGAG	CCCFFFFFHHHHHJJJJJJJJJJEHJJJJIIIJFEIJIHIIJIIIGIIHHEIIEGHGHGIFHIHIGIGIIJJHGHHHH;?CEEDDDDDDDD<-4>AC?B85	NM:i:0	MD:Z:101	AS:i:101	XS:i:0

  ```

---


# Filtering!

  * Tools : 
    * `pairsamtools` (https://github.com/mirnylab/pairsamtools) : filtering

  * Command
    ```
    samtools view -h <input_bam_file> | {
        pairsamtools parse -c <chromsize_file> 
    } | { 
        pairsamtools sort
    } | { 
        pairsamtools select '(pair_type == "CX") or (pair_type == "LL")'
    } | { 
        pairsamtools dedup --output \
            >( pairsamtools split \
                --output-pairs <output.pairs.gz> \
                --output-sam <output.bam> ) 
    }
    ```

  * input: `bam` file

  * output: `bam`, `pairs` file


---


# Pairs file
 
  * A `pairs` file is a block-gzipped text file that stores list of contact loci

  * Tools:
    * `pairix` (https://github.com/4dn-dcic/pairix) : indexing & querying `pairs` files

  * Command
    ```
    pairix <pairsfile> <querystr> [ <querystr> ... ] 
    ```
    ```
    pairix <pairsfile> 'chr1|chr1' 'chr2:1-2000000|chr4:5000000-6000000'
    ```

  * Pairs format: https://github.com/4dn-dcic/pairix/blob/master/pairs_format_specification.md

  * Example lines
  ```
  #columns: readID chr1 pos1 chr2 pos2 strand1 strand2
  EAS139:136:FC706VJ:2:2104:23462:197393 chr1 10000 chr1 20000 + +
  EAS139:136:FC706VJ:2:8762:23765:128766 chr1 50000 chr1 70000 + +
  EAS139:136:FC706VJ:2:2342:15343:9863 chr1 60000 chr2 10000 + +
  ```

---

# Binning!

  * Tools:
    * `cooler` (https://github.com/mirnylab/cooler)

  * Command
    ```
    cooler cload pairix <chrsize_file>:<bin_size> <pairs_file> <out_prefix>.cool
    ```

  * input: `pairs` file, `chromsize` file

  * output : `cool` file


---

# How to read a `cool` file

  * A `cool` file is an HDF5-format file that stores contact matrices and normailzation vectors.

  * Tools:
    * `cooler` (https://github.com/mirnylab/cooler)

  * Command
    ```
    cooler dump -b -t pixels --header --join -r chr3:1-12,000,000 -r2 chr17 out.cool 
    ```

  * Cool format : http://cooler.readthedocs.io/en/latest/datamodel.html


---

# Normalization! (Matrix balancing)

  * Tools:
    * `cooler` (https://github.com/mirnylab/cooler)

  * Command
    ```
    cooler balance <cool_file>
    ```

  * input : `cool` file
  
  * output : `cool` file


---

# Distiller 

  * A modular Hi-C mapping pipeline for reproducible data analysis.
  * Nextflow/Docker-based pipeline
  * https://github.com/mirnylab/distiller-nf


---

class: center, middle

# Visualization

To explore the features of a matrix, it helps to be able to see the data. Example with HiGlass:

<img src="img/sox2.png" width="600" />

---

class: center, middle

# Existing tools

<table style="margin: 0px auto">
    <tr>
        <td> Juicebox </td>
        <td> <img style="margin-left: 20px" src="img/juicebox-rectangle.png" height="120" /> </td>
    </tr>
    <tr>
        <td>WashU Epigenome Browser</td>
        <td><img style="margin-left: 20px" src="img/washu-epigenome-browser.png" height="120" /></td>
    </tr>
    <tr> 
        <td> 3D Genome Browser </td>
        <td><img style="margin-left: 20px" src="img/3d-genome-browser.png" height="120" /></td>
    </tr>
</table>

---

class: center, middle

# Task #1 

Re-create this figure:

<img src="https://cloud.githubusercontent.com/assets/2143629/24535688/a6043f1a-15a3-11e7-8197-d5d3ce227bc6.png" height=400/>

---

class: center, middle

### Navigate to HiGlass website

# [http://higlass.io/app](http://higlass.io/app)

---

class: center, middle

# 1. Remove genome position search bar

<video width="600" loop="loop" autoplay="autoplay">
    <source src="img/copy-replace-copy.mp4" type="video/mp4">
</video>

---

class: center, middle

# 2. Remove all existing tracks

<video width="620" loop="loop" autoplay="autoplay">
    <source src="img/remove-tracks.mp4" type="video/mp4">
</video>

---

class: center, middle

# 3. Add new dataset

<video width="620" loop="loop" autoplay="autoplay">
    <source src="img/add-schwarzer-wt.mp4" type="video/mp4">
</video>

---

class: center, middle

# 4. Resize the current view and replicate it

<video width="620" loop="loop" autoplay="autoplay">
    <source src="img/resize-and-triple.mp4" type="video/mp4">
</video>

---

class: center, middle

# 5. Replace existing dataset and copy it

<video width="620" loop="loop" autoplay="autoplay">
    <source src="img/replace-and-double.mp4" type="video/mp4">
</video>

---

class: center, middle

# 6. Lock zoom and location

<video width="620" loop="loop" autoplay="autoplay">
    <source src="img/lock-zoom.mp4" type="video/mp4">
</video>

---

# 7. Zoom top to: 

```
chr14:50,143,364-88,897,098 & chr14:70,582,768-97,882,196 [offset 0,0:0,0]
```

---

# 8 Zoom bottom to: 

```
chr14:66,384,132-76,643,845 & chr14:67,964,160-79,229,728 [offset 0,0:0,0]
```

---

# Exporting a link

---

# Running HiGlass locally

To look at private data, you have to run HiGlass locally. This is easily
accomplished using a docker container.

```
docker stop higlass-container; 
docker rm higlass-container;

docker pull gehlenborglab/higlass:v0.0.62 # higher versions are experimental and likely don't work


docker run --detach \
           --publish 8989:80 \
           --volume ~/hg-data:/data \
           --volume ~/tmp:/tmp \
           --name higlass-container \
           gehlenborglab/higlass:v0.0.62
```

Source: [https://github.com/hms-dbmi/higlass/wiki](https://github.com/hms-dbmi/higlass/wiki)

---

# Importing data

To load a local cooler file, place it in `~/tmp` and run the following command:

```
docker exec higlass-container python higlass-server/manage.py \
  ingest_tileset \
  --filename /tmp/matrix.multi.cool \
  --datatype matrix \
  --filetype cooler 
```

This makes the file accessible to `higlass` and registers it with the list of
available data. The datatype needs to be specified so that higlass knows what 
types of tracks can be used to display it. The filetype tells HiGlass how
it can be loaded.

Contrast with:

```
docker exec higlass-container python higlass-server/manage.py \
  ingest_tileset \
  --filename /tmp/file.hitile \
  --filetype hitile \
  --datatype vector \
  --name "Some 1D genomic data"
```

For vector bigWig data.

---


# Example pipeline

Starting from SRR1658602_{1,2}.fastq


    </textarea>
    <script src="https://remarkjs.com/downloads/remark-latest.min.js">
    </script>
    <script>
      var slideshow = remark.create();
    </script>
  </body>
</html>
