<!doctype html>
<html>
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

        <title>4DN Annual Meeting Tutorial</title>

        <link rel="stylesheet" href="css/reveal.css">
        <link rel="stylesheet" href="css/theme/presentation.css">

        <!-- Theme used for syntax highlighting of code -->
        <link rel="stylesheet" href="lib/css/zenburn.css">

        <!-- Printing and PDF exports -->
        <script>
var link = document.createElement( 'link' );
link.rel = 'stylesheet';
link.type = 'text/css';
link.href = window.location.search.match( /print-pdf/gi ) ? 'css/print/pdf.css' : 'css/print/paper.css';
document.getElementsByTagName( 'head' )[0].appendChild( link );
        </script>

    </head>
    <body>
        <div class="reveal">
            <div class="slides">
                <section>
                    <h3>DCIC Visualization Tools Tutorial</h3>

                    <h4>4DN Annual Meeting, Bethesda, MD</h4>

                    <i>Peter Kerpedjiev, Fritz Lekschas, Nils Gehlenborg</i>
                </section>
                
                <section>
                    <h3>WiFi Password</h3>

                    <b>Username:</b>
                    <b>Password:</b>
                </section>

                <section>
                    <h3> Existing tools </h3>

                    <p style="text-align: center">
                    <table style="margin: 0px auto">
                        <tr>
                            <td> Juicebox </td>
                            <td> <img style="margin-left: 20px" src="img/juicebox-rectangle.png" height="120" /> </td>
                        </tr>
                        <tr>
                            <td>WashU Epigenome Browser</td>
                            <td><img style="margin-left: 20px" src="img/washu-epigenome-browser.png" height="120" /></td>
                        </tr>
                        <tr> 
                            <td> 3D Genome Browser </td>
                            <td><img style="margin-left: 20px" src="img/3d-genome-browser.png" height="120" /></td>
                        </tr>
                    </table>
                    </p>
                </section>


                <section>
                    <h3> HiGlass </h3>


                    <p style="text-align: center">
                    <img src="img/higlass-overview.png" width=400 />
                    </p>

                    HiGlass is a viewer for HiC and other genomic data. It can be used
                    online (<a href="http://higlass.io/" >http://higlass.io</a>), or run locally with private data.
                </section>

                <section>
                    <h3> Capabilities </h3>

                    HiGlass is designed to facilitate the comparison of multiple genomic datasets through
                    "tracks" and "views".


                    <p style="text-align: center">
                    <img src="img/g12878_vs_K257.png" width=600/>
                    </p>
                </section>
                
                <section>
                    <h3> Anatomy of HiGlass </h3>

                    <table>
                        <tr>
                            <td>
                                <p style="text-align: center">
                                <img src="img/higlass-anatomy-11.png" width=400/>
                                </p>
                            </td>
                            <td>
                                <p style="margin-left: 15px">
                                HiGlass is composed of a collection of views and
                                tracks.
                                </p>
                            </td>
                        </tr>
                    </table>
                </section>

                <section>
                    <h3> Anatomy of HiGlass </h3>

                    <img src="img/higlass-anatomy-10.png" height=500/>

                    <p style="margin-left: 15px">
                    Each view is composed of a set of tracks which share
                    common axes. 
                    </p>
                </section>

                <section>
                    <h3> Anatomy of HiGlass </h3>

                    <img src="img/higlass-anatomy-9.png" height=500/>
                    <p style="margin-left: 15px">
                    Settings related to the view can be accessed via the 
                    "view settings" menu. 
                    </p>
                </section>

                <section>
                    <h3>Anatomy of HiGlass</h3>

                    <img src="img/higlass-anatomy-7.png" width=400/>
                    <p style="margin-left: 15px">
                    Views can be closed by clicking on the "close view" button ("X").
                    </p>
                </section>

                <section>
                    <h3>Anatomy of HiGlass</h3>

                <img src="img/higlass-anatomy-6.png" width=400/>
            <p style="margin-left: 15px">
            The contents of a view can be modified by adding tracks.
                </section>

                <section>
                    <h3>Anatomy of HiGlass</h3>
                <img src="img/higlass-anatomy-5.png" width=400/>
            <p style="margin-left: 15px">
            Tracks can be layed out horizontally where they track
            the horizontal axis.
            </p>
                </section>

                <section>
                    <h3>Anatomy of HiGlass</h3>
                    <img src="img/higlass-anatomy-4.png" width=400/>
                    <p style="margin-left: 15px">
                    Tracks can also be layed out vertically where they 
                    track the vertical axis.
                    </p>
                </section>

                <section>
                    <h3>Anatomy of HiGlass</h3>
                <img src="img/higlass-anatomy-3.png" width=400/>
            <p style="margin-left: 15px">
            And tracks can be placed in the center, where they move
            along both axes.
            </p>
                </section>

                <section>
                    <h3>Anatomy of HiGlass</h3>
                    <img src="img/higlass-anatomy-2.png" width=400/>
                    <p style="margin-left: 15px">
                    Track settings can be adjusted via the track settings
                    menu.
                    </p>
                </section>

                <section>
                    <h3>Anatomy of HiGlass</h3>
                <img src="img/higlass-anatomy-1.png" width=400/>
            <p style="margin-left: 15px">
                Overlays can be added on top of existing tracks by clicking
                on the '+' icon of the track toolbar.
            </p>
                </section>

                <section>
                    <h3>Anatomy of HiGlass</h3>
                <img src="img/higlass-anatomy-0.png" width=400/>
            <p style="margin-left: 15px">
                Tracks can be closed by clicking on the 'X' of the
                track toolbar.
            </p>
                </section>

                <section>
                    <h3>Practical Exercise</h3>

                    <p>Re-create this figure:</p>

<p style="text-align: center">
    <img src="https://cloud.githubusercontent.com/assets/2143629/24535688/a6043f1a-15a3-11e7-8197-d5d3ce227bc6.png" height=400/>
</p>
                </section>

                <section>

                    <h3> Navigate to HiGlass website</h3>

                    <a href="http://higlass.io">http://higlass.io/</a>
                    <p>And click on the "full-screen" icon</p>
                </section>

                <section>
                    <h3> Remove genome position search bar </h3>

<p style="text-align: center">
    <video width="600" loop="loop" autoplay="autoplay">
        <source src="img/remove-genome-position-search-bar.mp4" type="video/mp4">
    </video>
</p>

The genome position search bar serves to show the visible extent of the data.
It can also be used to navigate to a particular locus.
                </section>
                <section>

                    <h3> Remove all existing tracks</h3>

<p style="text-align: center">
    <video width="580" loop="loop" autoplay="autoplay">
        <source src="img/remove-tracks.mp4" type="video/mp4">
    </video>
</p>

We don't need any of the default tracks, so we'll remove them.
                </section>
                <section>

                    <h3> Add new dataset</h3>

<p style="text-align: center">
    <video width="620" loop="loop" autoplay="autoplay">
        <source src="img/add-schwarzer-wt.mp4" type="video/mp4">
    </video>
</p>

                </section>
                <section>

                    <h3> Resize the current view and replicate it </h3>

<p style="text-align: center">
    <video width="620" loop="loop" autoplay="autoplay">
        <source src="img/resize-and-triple.mp4" type="video/mp4">
    </video>
</p>
                </section>

                <section>

                    <h3> Replace existing dataset and copy it </h3>

<p style="text-align: center">
    <video width="560" loop="loop" autoplay="autoplay">
        <source src="img/replace-and-double.mp4" type="video/mp4">
    </video>
</p>
                </section>
                <section>

                    <h3> Lock zoom and location </h3>

<p style="text-align: center">
    <video width="560" loop="loop" autoplay="autoplay">
        <source src="img/lock-zoom.mp4" type="video/mp4">
    </video>
</p>

http://higlass.io/app/?config=FQWaiKoATPSMVYyOu11rBg
                </section>

                <section>
                    <h3> Zoom top to: </h3>

                    <pre>
chr14:50,143,364-88,897,098 & chr14:70,582,768-97,882,196 [offset 0,0:0,0]
                    </pre>

<p style="text-align: center">
    <video width="500" loop="loop" autoplay="autoplay">
        <source src="img/higlass-zoom-top.mp4" type="video/mp4">
    </video>
</p>

By providing a set of genomic coordinates, we can zoom to a defined region
in the heatmap.

</section>
<section>

    <h3> Zoom bottom to:</h3>

    <pre>
chr14:66,384,132-76,643,845 & chr14:67,964,160-79,229,728 [offset 0,0:0,0]
    </pre>

<p style="text-align: center">
    <video width="500" loop="loop" autoplay="autoplay">
        <source src="img/higlass-zoom-bottom.mp4" type="video/mp4">
    </video>
</p>

http://higlass.io/app/?config=JvtlSiVnSi-SSraD0zmsjA

</section>
<section>

    <h3> Project viewport </h3>

    <p style="text-align: center">
    <video width="600" loop="loop" autoplay="autoplay">
        <source src="img/project-viewport.mp4" type="video/mp4">
    </video>
    </p>

    Projecting a viewport shows the extent of one view on another. In this case,
    we're showing the extent of the bottom views on the top.
</section>
<section>

# Share a link


<p style="text-align: center">
    <video width="600" loop="loop" autoplay="autoplay">
        <source src="img/higlass-export-link.mp4" type="video/mp4">
    </video>
</p>

The current view configuration can be shared by exporting it as a link (http://higlass.io/app/?config=XVJAsgUbRva2diamiW9mfg)
</section>
<section>

    <h3> Add 1D track </h3>

<p style="text-align: center">
    <video width="580" loop="loop" autoplay="autoplay">
        <source src="img/higlass-add-rnaseq.mp4" type="video/mp4">
    </video>
</p>

We can add any sort of 1D data as a horizontal or vertical track. In this
case we add RNAseq data as horizontal track on top.
</section>
<section>

    <h3>Add gene annotations</h3>

<p style="text-align: center">
    <video width="580" loop="loop" autoplay="autoplay">
        <source src="img/higlass-add-annotations.mp4" type="video/mp4">
    </video>
</p>

Adding gene annotations lets us verify that the RNAseq profile matches
the locations of the genes (http://higlass.io/app/?config=YKBKin5SRhWIjuWwwnwpYw).
</section>
<section>

    <h3> Running HiGlass locally </h3>

To look at private data, you have to run HiGlass locally. This is easily
accomplished using a docker container.

The following commands can be found on the higlass wiki (google: higlass wiki) or go to: https://github.com/hms-dbmi/higlass/wiki

<pre>
docker stop higlass-container; 
docker rm higlass-container;

docker pull gehlenborglab/higlass:v0.2.6 


docker run --detach \
           --publish 8989:80 \
           --volume ~/hg-data:/data \
           --volume ~/tmp:/tmp \
           --name higlass-container \
           gehlenborglab/higlass:v0.2.6
           </pre>

</section>

<section>
    <h3> Importing data </h3>

We can import multi-res data into local instances of HiGlass. First, let's
get a multires file:

<pre>
wget https://s3.amazonaws.com/pkerp/public/matrix.multi.cool
</pre>

</section>
<section>

    <h3> Importing data </h3>

We can import multi-res data into local instances of HiGlass. First, let's
get a multires file:

```bash
wget https://s3.amazonaws.com/pkerp/public/matrix.multi.cool
```

To load it into HiGlass move it to `~/tmp`: 

```bash
cp matrix.multi.cool ~/tmp
```

</section>
<section>

    <h3> Importing data </h3>

We can import multi-res data into local instances of HiGlass. First, let's
get a multires file:

<pre>
wget https://s3.amazonaws.com/pkerp/public/matrix.multi.cool
</pre>

To load it into HiGlass move it to `~/tmp`: 

<pre>
cp matrix.multi.cool ~/tmp
</pre>

and run the following command:

<pre>
docker exec higlass-container python higlass-server/manage.py \
  ingest_tileset \
  --filename /tmp/matrix.multi.cool \
  --datatype matrix \
  --filetype cooler 
</pre>

This makes the file accessible to `higlass` and registers it with the list of
available data. 

</section>
<section>

    <h3> Displaying local data </h3>

<table>
    <tr>
        <td>
            <ul>
                <li style="margin-bottom: 20px">Navigate to <a href="http://localhost:8989">http://localhost:8989</a></li>
                <li style="margin-bottom: 20px">Click on the "Add track" button</li>
                <li style="margin-bottom: 20px">Search for "matrix"</li>
                <li style="margin-bottom: 20px">Double click on "matrix.multi.cool</li>
            </ul>
        </td>
        <td>
            <p style="text-align: center">
                <video width="500" loop="loop" autoplay="autoplay">
                    <source src="img/higlass-load-local.mp4" type="video/mp4">
                </video>
            </p>
        </td>
    </tr>
</table>

</section>
<section>

# Documentation:

 HiGlass Documentation
 <ul>
     <li> <a href="https://github.com/hms-dbmi/higlass/wiki">https://github.com/hms-dbmi/higlass/wiki</a>
     </li>
</ul>



Examples
<ul>
    <li> <a href="http://higlass.io/examples">http://higlass.io/examples</a>
    </li>
</ul>

</section>

                <section>
                    <h3>More Information</h3>
                    Posters #31 and #41 tonight!
                </section>
            </div>
        </div>

        <script src="lib/js/head.min.js"></script>
        <script src="js/reveal.js"></script>

        <script>
// More info about config & dependencies:
// - https://github.com/hakimel/reveal.js#configuration
// - https://github.com/hakimel/reveal.js#dependencies
Reveal.initialize({
    height: 800,
    history: true,
    dependencies: [
        { src: 'plugin/markdown/marked.js' },
        { src: 'plugin/markdown/markdown.js' },
        { src: 'plugin/notes/notes.js', async: true },
        { src: 'plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } }
    ],

    minScale: 1,
    maxScale: 1
});
        </script>
    </body>

</html>
