{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-19T10:14:59.124718Z",
     "start_time": "2017-07-19T10:14:58.793436Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, '/net/proteome/home/nezar/local/devel/bioframe/')\n",
    "sys.path.insert(0, '/net/proteome/home/nezar/local/devel/mirnymental/nezar/')\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-19T10:15:01.108850Z",
     "start_time": "2017-07-19T10:14:59.126388Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from matplotlib.gridspec import GridSpec\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import seaborn.apionly as sns\n",
    "import mirnylib.plotting\n",
    "mpl.style.use('seaborn-poster')\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import bioframe\n",
    "import bioframe.dask\n",
    "import dask\n",
    "import distributed\n",
    "import cooler\n",
    "from scipy.linalg import toeplitz\n",
    "import cooler_expected_new as cooler_expected\n",
    "\n",
    "from cooler.contrib.dask import daskify"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-19T10:15:01.125890Z",
     "start_time": "2017-07-19T10:15:01.110785Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from __future__ import division, print_function\n",
    "from multiprocess import Pool\n",
    "from functools import partial\n",
    "from itertools import chain\n",
    "import itertools\n",
    "import argparse\n",
    "import sys\n",
    "\n",
    "from scipy.signal import fftconvolve\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import cooler\n",
    "from cooler.contrib.dask import daskify\n",
    "from cooler.tools import split, partition\n",
    "import dask.dataframe as dd\n",
    "import dask.array as da\n",
    "\n",
    "where = np.flatnonzero\n",
    "concat = chain.from_iterable\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-19T14:56:30.040749Z",
     "start_time": "2017-07-19T14:56:28.960841Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def lattice_pdist_frequencies(n, points):\n",
    "    \"\"\"\n",
    "    Distribution of pairwise 1D distances among a collection of distinct\n",
    "    integers ranging from 0 to n-1.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    n : int\n",
    "        Size of the lattice on which the integer points reside.\n",
    "    points : sequence of int\n",
    "        Arbitrary integers between 0 and n-1, inclusive, in any order but \n",
    "        with no duplicates.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    h : 1D array of length n\n",
    "        h[d] counts the number of integer pairs that are exactly d units apart\n",
    "\n",
    "    Notes\n",
    "    -----\n",
    "    This is done using a convolution via FFT. Thanks to Peter de Rivaz; see\n",
    "    <http://stackoverflow.com/questions/42423823/distribution-of-pairwise-\n",
    "    distances-between-many-integers>.\n",
    "    \n",
    "    \"\"\"\n",
    "    if len(np.unique(points)) != len(points):\n",
    "        raise ValueError(\"Integers must be distinct.\")\n",
    "    x = np.zeros(n); x[points] = 1\n",
    "    return np.round(fftconvolve(x, x[::-1], mode='full')).astype(int)[-n:]\n",
    "\n",
    "\n",
    "def count_bad_pixels_per_diag(n, bad_bins):\n",
    "    \"\"\"\n",
    "    Efficiently count the number of bad pixels on each upper diagonal of a\n",
    "    matrix assuming a sequence of bad bins forms a \"grid\" of invalid pixels.\n",
    "    \n",
    "    Each bad bin bifurcates into two a row and column of bad pixels, so an \n",
    "    upper bound on number of bad pixels per diagonal is 2*k, where k is the \n",
    "    number of bad bins. For a given diagonal, we need to subtract from this\n",
    "    upper estimate the contribution from rows/columns reaching \"out-of-bounds\"\n",
    "    and the contribution of the intersection points of bad rows with bad\n",
    "    columns that get double counted.\n",
    "    \n",
    "    o : bad bin\n",
    "    * : bad pixel\n",
    "    x : intersection bad pixel\n",
    "    $ : out of bounds bad pixel\n",
    "\n",
    "         $    $     $\n",
    "     *--------------------------+    \n",
    "      *  *    *     *           |    \n",
    "       * *    *     *           |    \n",
    "        **    *     *           |    \n",
    "         o****x*****x***********|$    \n",
    "          *   *     *           |    \n",
    "           *  *     *           |    \n",
    "            * *     *           |    \n",
    "             o******x***********|$    \n",
    "              *     *           |    \n",
    "               *    *           |    \n",
    "                *   *           |    \n",
    "                 *  *           |    \n",
    "                  * *           |    \n",
    "                   **           |    \n",
    "                    o***********|$    \n",
    "                     *          |    \n",
    "                      *         |    \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    n : int\n",
    "        total number of bins\n",
    "    bad_bins : 1D array of int\n",
    "        sorted array of bad bin indexes\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    dcount : 1D array of length n\n",
    "        dcount[d] == number of bad pixels on diagonal d\n",
    "    \n",
    "    \"\"\"\n",
    "    k = len(bad_bins)\n",
    "    dcount = np.zeros(n, dtype=int)\n",
    "\n",
    "    # Store all intersection pixels in a separate array\n",
    "    # ~O(n log n) with fft\n",
    "    ixn = lattice_pdist_frequencies(n, bad_bins)\n",
    "    dcount[0] = ixn[0]\n",
    "\n",
    "    # Keep track of out-of-bounds pixels by squeezing left and right bounds\n",
    "    # ~O(n)\n",
    "    pl = 0\n",
    "    pr = k\n",
    "    for diag in range(1, n):\n",
    "        if pl < k:\n",
    "            while (bad_bins[pl] - diag) < 0:\n",
    "                pl += 1\n",
    "                if pl == k:\n",
    "                    break\n",
    "        if pr > 0:\n",
    "            while (bad_bins[pr-1] + diag) >= n:\n",
    "                pr -= 1\n",
    "                if pr == 0:\n",
    "                    break\n",
    "        dcount[diag] = 2*k - ixn[diag] - pl - (k - pr)\n",
    "    return dcount\n",
    "\n",
    "\n",
    "def count_all_pixels_per_diag(n):\n",
    "    \"\"\"\n",
    "    Total number of pixels on each upper diagonal of a square matrix.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    n : int\n",
    "        total number of bins (dimension of square matrix)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    dcount : 1D array of length n\n",
    "        dcount[d] == total number of pixels on diagonal d\n",
    "\n",
    "    \"\"\"\n",
    "    return np.arange(n, 0, -1)\n",
    "\n",
    "\n",
    "\n",
    "def make_diag_table(bad_mask, span1, span2):\n",
    "    \"\"\"\n",
    "    Compute the total number of elements ``n_elem`` and the number of bad \n",
    "    elements ``n_bad`` per diagonal for a single contact area encompassing \n",
    "    ``span1`` and ``span2`` on the same genomic scaffold (cis matrix).\n",
    "\n",
    "    Follows the same principle as the algorithm for finding contact areas for\n",
    "    computing scalings.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    bad_mask : 1D array of bool\n",
    "        Mask of bad bins for the whole genomic scaffold containing the regions\n",
    "        of interest.\n",
    "    span1, span2 : pair of ints\n",
    "        The bin spans (not genomic coordinates) of the two regions of interest.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    diags : pandas.DataFrame\n",
    "        Table indexed by 'diag' with columns ['n_elem', 'n_bad'].\n",
    "\n",
    "    \"\"\"\n",
    "    def _make_diag_table(n_bins, bad_locs):\n",
    "        diags = pd.DataFrame(index=pd.Series(np.arange(n_bins), name='diag'))\n",
    "        diags['n_valid'] = (count_all_pixels_per_diag(n_bins) - \n",
    "                            count_bad_pixels_per_diag(n_bins, bad_locs))\n",
    "        return diags\n",
    "    \n",
    "    if span1 == span2:\n",
    "        lo, hi = span1\n",
    "        diags = _make_diag_table(hi - lo, where(bad_mask[lo:hi]))\n",
    "    else:\n",
    "        lo1, hi1 = span1\n",
    "        lo2, hi2 = span2\n",
    "        if lo2 <= lo1:\n",
    "            lo1, lo2 = lo2, lo1\n",
    "            hi1, hi2 = hi2, hi1\n",
    "        diags = (\n",
    "            _make_diag_table(hi2 - lo1, where(bad_mask[lo1:hi2]))\n",
    "                .subtract(_make_diag_table(lo2 - lo1, where(bad_mask[lo1:lo2])), \n",
    "                          fill_value=0)\n",
    "                .subtract(_make_diag_table(hi2 - hi1, where(bad_mask[hi1:hi2])), \n",
    "                          fill_value=0)            \n",
    "        )\n",
    "        if hi1 < lo2:\n",
    "            diags.add(_make_diag_table(lo2 - hi1, where(bad_mask[hi1:lo2])), \n",
    "                      fill_value=0)\n",
    "        diags = diags[diags['n_elem'] > 0]\n",
    "    return diags.astype(int)\n",
    "\n",
    "\n",
    "def sum_diagonals(df, field):\n",
    "    reduced = df.groupby('diag')[field].sum()\n",
    "    reduced.name = field + '.sum'\n",
    "    return reduced\n",
    "\n",
    "\n",
    "def compute_expected(clr, regions, field='balanced', chunksize=1000000):\n",
    "    pixels = daskify(clr.filename, 'pixels', chunksize=chunksize)\n",
    "    pixels = cooler.annotate(pixels, clr.bins(), replace=False)\n",
    "    pixels = pixels[pixels.chrom1 == pixels.chrom2]\n",
    "    chroms = [region[0] for region in regions]\n",
    "    cis_maps = {chrom: pixels[pixels.chrom1==chrom] for chrom in chroms}\n",
    "\n",
    "    diag_tables = []\n",
    "    data_sums = []\n",
    "\n",
    "    for region in regions:\n",
    "        if len(region) == 1:\n",
    "            chrom, = region\n",
    "            start1, end1 = 0, clr.chromsizes[chrom]\n",
    "            start2, end2 = start1, end1\n",
    "        elif len(region) == 3:\n",
    "            chrom, start1, end1 = region\n",
    "            start2, end2 = start1, end1\n",
    "        elif len(region) == 5:\n",
    "            chrom, start1, end1, start2, end2 = region\n",
    "        else:\n",
    "            raise ValueError(\"Regions must be sequences of length 1, 3 or 5\")\n",
    "\n",
    "        bins = clr.bins().fetch(chrom).reset_index(drop=True)\n",
    "        bad_mask = np.array(bins['weight'].isnull())\n",
    "        lo1, hi1 = clr.extent((chrom, start1, end1))\n",
    "        lo2, hi2 = clr.extent((chrom, start2, end2))\n",
    "        co = clr.offset(chrom)\n",
    "        lo1 -= co\n",
    "        lo2 -= co\n",
    "        hi1 -= co\n",
    "        hi2 -= co\n",
    "\n",
    "        dt = make_diag_table(bad_mask, [lo1, hi1], [lo2, hi2])\n",
    "        sel = bioframe.bg2slice_frame(\n",
    "            cis_maps[chrom], \n",
    "            (chrom, start1, end1), \n",
    "            (chrom, start2, end2)\n",
    "        ).copy()\n",
    "        sel['diag'] = sel['bin2_id'] - sel['bin1_id']\n",
    "        sel['balanced'] = sel['count'] * sel['weight1'] * sel['weight2']\n",
    "        agg = sum_diagonals(sel, field)\n",
    "        diag_tables.append(dt)\n",
    "        data_sums.append(agg)\n",
    "\n",
    "    # run dask scheduler\n",
    "    if len(data_sums) and isinstance(data_sums[0], dd.Series):\n",
    "        data_sums = dd.compute(*data_sums)\n",
    "\n",
    "    # append to tables\n",
    "    for dt, agg in zip(diag_tables, data_sums):\n",
    "        dt[agg.name] = 0\n",
    "        dt[agg.name] = dt[agg.name].add(agg, fill_value=0)\n",
    "        dt.loc[:2, agg.name] = np.nan\n",
    "\n",
    "    # merge and return\n",
    "    dtable = pd.concat(diag_tables, \n",
    "                     keys=list(chroms),\n",
    "                     names=['chrom'])\n",
    "    dtable['balanced.avg'] = dtable['balanced.sum'] / dtable['n_valid']\n",
    "    return dtable\n",
    "\n",
    "\n",
    "def compute_trans_expected(clr, pixels, chromosomes):\n",
    "    def n_total_trans_elements(clr, chromosomes):\n",
    "        n = 0\n",
    "        x = [clr.extent(chrom)[1] - clr.extent(chrom)[0] \n",
    "                 for chrom in chromosomes]\n",
    "        for i in range(len(x)):\n",
    "            for j in range(i + 1, len(x)):\n",
    "                n += x[i] * x[j]\n",
    "        return n\n",
    "\n",
    "    def n_bad_trans_elements(clr, chromosomes):\n",
    "        n = 0\n",
    "        x = [np.sum(clr.bins()['weight']\n",
    "                       .fetch(chrom)\n",
    "                       .isnull()\n",
    "                       .astype(int)\n",
    "                       .values)\n",
    "                 for chrom in chromosomes]\n",
    "        for i in range(len(x)):\n",
    "            for j in range(i + 1, len(x)):\n",
    "                n += x[i] * x[j]\n",
    "        return n\n",
    "\n",
    "    pixels = cooler.annotate(pixels, clr.bins(), replace=False)\n",
    "    pixels = pixels[\n",
    "        (pixels.chrom1.isin(chromosomes)) &\n",
    "        (pixels.chrom2.isin(chromosomes)) &\n",
    "        (pixels.chrom1 != pixels.chrom2)\n",
    "    ]\n",
    "    pixels['balanced'] = pixels['count'] * pixels['weight1'] * pixels['weight2']\n",
    "    ntot = n_total_trans_elements(clr, chromosomes)\n",
    "    nbad = n_bad_trans_elements(clr, chromosomes)\n",
    "    trans_area = ntot - nbad\n",
    "    trans_sum = pixels['balanced'].sum().compute()\n",
    "    return trans_sum, trans_area\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-19T14:55:40.969337Z",
     "start_time": "2017-07-19T14:55:39.502312Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Client: scheduler='tcp://127.0.0.1:33799' processes=8 cores=40>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster = distributed.LocalCluster(n_workers=8)\n",
    "client = distributed.Client(cluster)\n",
    "client"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-19T14:55:43.890139Z",
     "start_time": "2017-07-19T14:55:43.847282Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "binsize = 5000\n",
    "longnames = {\n",
    "    'U': 'UNTR',\n",
    "    'T': 'TAM',\n",
    "    'N': 'NIPBL'\n",
    "}\n",
    "clrs = {\n",
    "    'U': cooler.Cooler(f'/net/levsha/share/nezar/TCCSpitz2015_distiller/output/coolers/library_group/UNTR.{binsize}.cool'),\n",
    "    'T': cooler.Cooler(f'/net/levsha/share/nezar/TCCSpitz2015_distiller/output/coolers/library_group/TAM.{binsize}.cool'),\n",
    "    'N': cooler.Cooler(f'/net/levsha/share/nezar/TCCSpitz2015_distiller/output/coolers/library_group/NIPBL.{binsize}.cool'),\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-19T14:55:44.929650Z",
     "start_time": "2017-07-19T14:55:44.711152Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "chromsizes = cooler.util.fetch_chromsizes('mm9')[:'chr19']\n",
    "chromosomes = list(chromsizes.index)\n",
    "regions = [(chrom, 0, chromsizes[chrom]) for chrom in chromosomes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-19T14:57:55.879151Z",
     "start_time": "2017-07-19T14:56:31.971693Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "expected = {}\n",
    "for cond in ['U', 'T', 'N']:\n",
    "    expected[cond] = compute_expected(clrs[cond], regions)\n",
    "    expected[cond].to_csv('{}.{}.expected.txt'.format(longnames[cond], binsize))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-19T14:57:57.563291Z",
     "start_time": "2017-07-19T14:57:57.509634Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>n_valid</th>\n",
       "      <th>balanced.sum</th>\n",
       "      <th>balanced.avg</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrom</th>\n",
       "      <th>diag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">chr1</th>\n",
       "      <th>0</th>\n",
       "      <td>37505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>36765</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>36719</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>36712</td>\n",
       "      <td>347.257539</td>\n",
       "      <td>0.009459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>36706</td>\n",
       "      <td>277.175077</td>\n",
       "      <td>0.007551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>36696</td>\n",
       "      <td>227.211984</td>\n",
       "      <td>0.006192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>36692</td>\n",
       "      <td>193.572324</td>\n",
       "      <td>0.005276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>36682</td>\n",
       "      <td>166.844372</td>\n",
       "      <td>0.004548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>36682</td>\n",
       "      <td>147.194655</td>\n",
       "      <td>0.004013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>36670</td>\n",
       "      <td>133.408979</td>\n",
       "      <td>0.003638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>36668</td>\n",
       "      <td>124.061074</td>\n",
       "      <td>0.003383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>36664</td>\n",
       "      <td>110.899576</td>\n",
       "      <td>0.003025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>36664</td>\n",
       "      <td>103.791855</td>\n",
       "      <td>0.002831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>36659</td>\n",
       "      <td>93.872428</td>\n",
       "      <td>0.002561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>36649</td>\n",
       "      <td>88.404068</td>\n",
       "      <td>0.002412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>36652</td>\n",
       "      <td>84.594194</td>\n",
       "      <td>0.002308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>36645</td>\n",
       "      <td>79.354078</td>\n",
       "      <td>0.002165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>36638</td>\n",
       "      <td>76.443009</td>\n",
       "      <td>0.002086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>36642</td>\n",
       "      <td>69.458462</td>\n",
       "      <td>0.001896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>36638</td>\n",
       "      <td>69.667872</td>\n",
       "      <td>0.001902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>36628</td>\n",
       "      <td>66.086740</td>\n",
       "      <td>0.001804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>36624</td>\n",
       "      <td>61.828751</td>\n",
       "      <td>0.001688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>36620</td>\n",
       "      <td>60.540031</td>\n",
       "      <td>0.001653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>36617</td>\n",
       "      <td>57.862785</td>\n",
       "      <td>0.001580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>36617</td>\n",
       "      <td>58.010012</td>\n",
       "      <td>0.001584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>36610</td>\n",
       "      <td>52.665356</td>\n",
       "      <td>0.001439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>36615</td>\n",
       "      <td>53.062673</td>\n",
       "      <td>0.001449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>36605</td>\n",
       "      <td>51.765142</td>\n",
       "      <td>0.001414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>36600</td>\n",
       "      <td>48.867265</td>\n",
       "      <td>0.001335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>36596</td>\n",
       "      <td>48.526980</td>\n",
       "      <td>0.001326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">chr19</th>\n",
       "      <th>12239</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12240</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12241</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12242</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12243</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12244</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12245</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12246</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12247</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12248</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12249</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12250</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12251</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12252</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12253</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12254</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12255</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12256</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12257</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12258</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12259</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12260</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12261</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12262</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12263</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12264</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12265</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12266</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12267</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12268</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>494476 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             n_valid  balanced.sum  balanced.avg\n",
       "chrom diag                                      \n",
       "chr1  0        37505           NaN           NaN\n",
       "      1        36765           NaN           NaN\n",
       "      2        36719           NaN           NaN\n",
       "      3        36712    347.257539      0.009459\n",
       "      4        36706    277.175077      0.007551\n",
       "      5        36696    227.211984      0.006192\n",
       "      6        36692    193.572324      0.005276\n",
       "      7        36682    166.844372      0.004548\n",
       "      8        36682    147.194655      0.004013\n",
       "      9        36670    133.408979      0.003638\n",
       "      10       36668    124.061074      0.003383\n",
       "      11       36664    110.899576      0.003025\n",
       "      12       36664    103.791855      0.002831\n",
       "      13       36659     93.872428      0.002561\n",
       "      14       36649     88.404068      0.002412\n",
       "      15       36652     84.594194      0.002308\n",
       "      16       36645     79.354078      0.002165\n",
       "      17       36638     76.443009      0.002086\n",
       "      18       36642     69.458462      0.001896\n",
       "      19       36638     69.667872      0.001902\n",
       "      20       36628     66.086740      0.001804\n",
       "      21       36624     61.828751      0.001688\n",
       "      22       36620     60.540031      0.001653\n",
       "      23       36617     57.862785      0.001580\n",
       "      24       36617     58.010012      0.001584\n",
       "      25       36610     52.665356      0.001439\n",
       "      26       36615     53.062673      0.001449\n",
       "      27       36605     51.765142      0.001414\n",
       "      28       36600     48.867265      0.001335\n",
       "      29       36596     48.526980      0.001326\n",
       "...              ...           ...           ...\n",
       "chr19 12239        0      0.000000           NaN\n",
       "      12240        0      0.000000           NaN\n",
       "      12241        0      0.000000           NaN\n",
       "      12242        0      0.000000           NaN\n",
       "      12243        0      0.000000           NaN\n",
       "      12244        0      0.000000           NaN\n",
       "      12245        0      0.000000           NaN\n",
       "      12246        0      0.000000           NaN\n",
       "      12247        0      0.000000           NaN\n",
       "      12248        0      0.000000           NaN\n",
       "      12249        0      0.000000           NaN\n",
       "      12250        0      0.000000           NaN\n",
       "      12251        0      0.000000           NaN\n",
       "      12252        0      0.000000           NaN\n",
       "      12253        0      0.000000           NaN\n",
       "      12254        0      0.000000           NaN\n",
       "      12255        0      0.000000           NaN\n",
       "      12256        0      0.000000           NaN\n",
       "      12257        0      0.000000           NaN\n",
       "      12258        0      0.000000           NaN\n",
       "      12259        0      0.000000           NaN\n",
       "      12260        0      0.000000           NaN\n",
       "      12261        0      0.000000           NaN\n",
       "      12262        0      0.000000           NaN\n",
       "      12263        0      0.000000           NaN\n",
       "      12264        0      0.000000           NaN\n",
       "      12265        0      0.000000           NaN\n",
       "      12266        0      0.000000           NaN\n",
       "      12267        0      0.000000           NaN\n",
       "      12268        0      0.000000           NaN\n",
       "\n",
       "[494476 rows x 3 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expected['U']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-07-19T10:17:46.750840Z",
     "start_time": "2017-07-19T10:17:44.514961Z"
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'X' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-3029d2f8fd4e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m     \u001b[0max\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m     \u001b[0mimg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmatshow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlog10\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcmap\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'fall'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvmin\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvmax\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1.5\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m     \u001b[0mcax\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m     \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolorbar\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mimg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcax\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcax\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0morientation\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'horizontal'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'X' is not defined"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAATkAAAEdCAYAAAB33BDRAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAE1ZJREFUeJzt3X+wZ3V93/Hni10Ud+0MMNbMJIFddqS4y0QprAkjJl22\nsZAmXdoxiRk7BjUStOOvOmrIaLftSlMGLDrGaUCGUUvISCWkkEQDkR+mMqEJawOyOMLqrksyWFQW\nkB+iW9/945zbXK737vd8v99779797PMxc+bc+/mezzmfz73f85rPOd/zPSdVhSS16qhD3QBJWkqG\nnKSmGXKSmmbISWqaISepaYacpKYNCrkkP5nkd5P8ZZKnk1SS9QPrHpPksiQPJ3mmX8fPTdNoSRpq\n6EjuJcCvAvuB/znmNq4GLgC2A78EPAzcnOS0MdcjSWPLkIuBkxxVVT/sf34zcBVwUlXtHVHv5cDf\nAG+qqk/0ZauBXcBXq2rbdM2XpIMbNJKbCbgJbAN+AFw3a10HgE8D5yR5/oTrlaRBlvqDh1OBPVX1\n9JzyXcDz6A6DJWnJrF7i9R9Pdx5vrkdnvT6WJDtnfq6qMyZsl6QjxFKH3JJ60YteVOvXrz/UzZC0\niHbu3PntqvqHi7W+pQ65/cC6ecpnRnCPzvPaQc0evW3evLnuvvvuCZsmaSVK8o3FXN9Sn5PbBZyU\nZM2c8k3A94HdS7x9SUe4pQ65PwaOBn5lpqC/hOS1wC1V9ewSb1/SEW7w4WqSX+5/nDlc/IUk3wK+\nVVVfSLIO+Bqwo6p2AFTV/05yHfCRJEcDe4C3AicB/3qxOiFJCxnnnNxn5vz+X/v5F4AtQIBV/Ojo\n8I3AfwIuBo4F7gHOraovjdtYSRrX4JCrqox4fS9d0M0tfwZ4dz9J0rLyLiSSmmbISWqaISepaYac\npKYZcpKaZshJapohJ6lphpykphlykppmyElqmiEnqWmGnKSmGXKSmmbISWqaISepaYacpKYZcpKa\nZshJapohJ6lphpykphlykppmyElqmiEnqWmGnKSmGXKSmmbISWqaISepaYacpKYZcpKaZshJapoh\nJ6lphpykphlykppmyElq2qCQS3JCkuuTPJ7kiSQ3JDlxYN0Tk3wqyb4kzyR5IMnFSdZO13RJGm31\nqAWSrAFuA54FzgcKuBi4PcnLquqpg9RdC3weOBr4d8A+4BXAfwROBl47bQck6WBGhhxwAbABOKWq\ndgMkuRd4ELgQuPwgdc+iC7Nzq+rmvuz2JMcD70mypqqenrj1kjTCkMPVbcBdMwEHUFV7gDuB80bU\nfV4/f2xO+WP9tjOwnZI0kSEhdypw3zzlu4BNI+p+nm7Ed2mSTUlemGQr8E7gioMd6krSYhgScscD\n++cpfxQ47mAVq+p7wKv67ewCvgvcCvwJ8LaxWtpLsnNmmqS+pCPLkHNyE0tyDHAd8GPA6+k+ePhp\nYDtwAHjrUm5fkoaE3H7mH7EtNMKb7TeALcDJs87p/UWSx4GPJ7miqu4Z2liAqjpj5ufNmzfXOHUl\nHXmGHK7uojsvN9cm4P4RdX8KeGz2hxa9v+rnGwdsX5ImNiTkbgLOTLJhpiDJerrLQ24aUfebwLFJ\nXjKn/Gf6+d8Na6YkTWZIyF0F7AVuTHJekm3AjcBDwJUzCyVZl+RAku2z6n6S7sOGzyY5P8nZSd4L\nfAjYSXcZiiQtmZEh11/msRV4ALgGuBbYA2ytqidnLRpg1ex1VtVe4Ezgb+i+JfFZuouLPw68uqp+\nuCi9kKQFDPp0tar2Aa8Zscxe5rm4t6ruB351ksZJ0rS8C4mkphlykppmyElqmiEnqWmGnKSmGXKS\nmmbISWqaISepaYacpKYZcpKaZshJapohJ6lphpykphlykppmyElqmiEnqWmGnKSmGXKSmmbISWqa\nISepaYacpKYZcpKaZshJapohJ6lphpykphlykppmyElqmiEnqWmGnKSmGXKSmmbISWqaISepaYac\npKYZcpKaNijkkpyQ5Pokjyd5IskNSU4cupEkG5N8Jsm3kzyT5KtJ3jl5syVpmNWjFkiyBrgNeBY4\nHyjgYuD2JC+rqqdG1N/c178DeDPwOHAy8MKpWi5JA4wMOeACYANwSlXtBkhyL/AgcCFw+UIVkxwF\n/Dfg1qr6V7Neun3iFkvSGIYcrm4D7poJOICq2gPcCZw3ou4WYCMHCUJJWkpDQu5U4L55yncBm0bU\nfVU/PybJXUl+kOSRJB9N8oJxGipJkxgScscD++cpfxQ4bkTdH+/n1wG3AK8GLqU7N/cHA9v4HEl2\nzkyT1Jd0ZBlyTm4aMyH6+1W1vf/5jiSrgEuSbKyqryxxGyQdwYaM5PYz/4htoRHebN/p538+p/yW\nfn7agO0/R1WdMTONW1fSkWdIyO2iOy831ybg/gF1JemQGRJyNwFnJtkwU5BkPXBW/9rBfI7u+rpz\n5pSf28//elArJWlCQ0LuKmAvcGOS85JsA24EHgKunFkoybokB5LMnHujqr4D/GfgLUl+J8nPJ7kI\n2A58avZlKZK0FEZ+8FBVTyXZCnwYuAYIcCvwrqp6ctaiAVbxo8G5A/gu8G+A9wAPA5cBH5y69ZI0\nwqBPV6tqH/CaEcvspQu6ueVFdzGwFwRLWnbehURS0ww5SU0z5CQ1zZCT1DRDTlLTDDlJTTPkJDXN\nkJPUNENOUtMMOUlNM+QkNc2Qk9Q0Q05S0ww5SU0z5CQ1zZCT1DRDTlLTDDlJTTPkJDXNkJPUNENO\nUtMMOUlNM+QkNc2Qk9Q0Q05S0ww5SU0z5CQ1zZCT1DRDTlLTDDlJTTPkJDXNkJPUNENOUtMMOUlN\nGxRySU5Icn2Sx5M8keSGJCeOu7EkFyWpJF8cv6mSNL6RIZdkDXAb8FLgfOD1wMnA7UnWDt1Qkg3A\nB4BHJmuqJI1v9YBlLgA2AKdU1W6AJPcCDwIXApcP3NbvAdcCpwzcriRNbcjh6jbgrpmAA6iqPcCd\nwHlDNpLkdcDpwG9P0khJmtSQkDsVuG+e8l3AplGVkxwHfBh4X1U9Ol7zJGk6Q0LueGD/POWPAscN\nqH8Z8ADwyeHNWliSnTPTYqxPUtuW9NxYkp8Ffh04vapqKbclSfMZMpLbz/wjtoVGeLNdCVwN/G2S\nY5McSxesq/rfnz9Wa4GqOmNmGreupCPPkJHcLrrzcnNtAu4fUXdjP71lntf2A/8W+MiANkjSRIaE\n3E3Ah5JsqKqvAyRZD5wFXDSi7tnzlH0EWAW8Hdg9z+uStGiGhNxVwNuAG5N8ACjgg8BDdIejACRZ\nB3wN2FFVOwCq6o65K0vyGLB6vtckabGNPCdXVU8BW+k+Ib2G7oLePcDWqnpy1qKhG6H5fVhJK8ag\nT1erah/wmhHL7KULulHr2jJkm5K0GBx1SWqaISepaYacpKYZcpKaZshJapohJ6lphpykphlykppm\nyElqmiEnqWmGnKSmGXKSmmbISWqaISepaYacpKYZcpKaZshJapohJ6lphpykphlykppmyElqmiEn\nqWmGnKSmGXKSmmbISWqaISepaYacpKYZcpKaZshJapohJ6lphpykphlykppmyElqmiEnqWmDQi7J\nCUmuT/J4kieS3JDkxAH1XpHk6iQPJnk6yb4k1yY5afqmS9JoI0MuyRrgNuClwPnA64GTgduTrB1R\n/bXAqcBHgX8OXAScDtyd5IQp2i1Jg6wesMwFwAbglKraDZDkXuBB4ELg8oPUvbSq3jO7IMmdwJ5+\nvdsnabQkDTXkcHUbcNdMwAFU1R7gTuC8g1WsqkfmKfsG8C3gJ8ZrqiSNb0jInQrcN0/5LmDTuBtM\nshF4MfCVcetK0riGhNzxwP55yh8FjhtnY0lWA1fQjeSuHqfurHXsnJkmqS/pyDLknNxi+hjwSuAX\nq2q+4JSkRTVkJLef+UdsC43w5pXkEuA3gTdV1S1D681VVWfMTJOuQ9KRY8hIbhfdebm5NgH3D9lI\nkvcDvwW8vaquGd48SZrOkJHcTcCZSTbMFCRZD5zVv3ZQSd4BXAy8v6o+NlkzJWkyQ0LuKmAvcGOS\n85JsA24EHgKunFkoybokB5Jsn1X2a8BHgD8Dbkty5qxp7E9mJWlcIw9Xq+qpJFuBDwPXAAFuBd5V\nVU/OWjTAKp4bnOf25ef202xfALZM3HJJGmDQp6tVtQ94zYhl9tIF2uyyNwBvmKxpkjQ970IiqWmG\nnKSmGXKSmmbISWqaISepaYacpKYZcpKaZshJapohJ6lphpykphlykppmyElqmiEnqWmGnKSmGXKS\nmmbISWqaISepaYacpKYZcpKaZshJapohJ6lphpykphlykppmyElqmiEnqWmGnKSmGXKSmmbISWqa\nISepaYacpKYZcpKaZshJapohJ6lphpykphlykpo2KOSSnJDk+iSPJ3kiyQ1JThxY95gklyV5OMkz\nSf4yyc9N12xJGmZkyCVZA9wGvBQ4H3g9cDJwe5K1A7ZxNXABsB34JeBh4OYkp03aaEkaavWAZS4A\nNgCnVNVugCT3Ag8CFwKXL1QxycuB1wFvqqpP9GVfAHYBO4BtU7VekkYYcri6DbhrJuAAqmoPcCdw\n3oC6PwCum1X3APBp4Jwkzx+7xZI0hiEhdypw3zzlu4BNA+ruqaqn56n7POAlA7b/HEl2zkzj1pV0\n5BlyuHo8sH+e8keB46aoO/P6pF66c+fOA0numWIdK8HGfv6VQ9qK6diHlaGVPrx8MVc4JORWlKo6\nA7oRXf/75kPboum00A/7sDK01IfFNORwdT/zj9gWGqUNrQt/P6KTpCUxJOR20Z1bm2sTcP+Auif1\nl6HMrft9YPePVpGkxZOqOvgCybuADwH/qKq+3petp7uE5KKq+i8HqfuPgS8Bb6iqT/Vlq4EvA7ur\n6l8sQh8kaUFDQm4tcA/wDPABoIAPAv8AeFlVPdkvtw74GrCjqnbMqv9p4BzgvcAe4K10FwW/sqq+\ntNgdkqTZRh6uVtVTwFbgAeAa4Fq6sNo6E3C9AKvmWecbgU8AFwN/CpwAnGvASVoOI0dyknQ48y4k\nkppmyElqmiEnqWmGnKSmGXKSmmbISWqaISepaSsy5Fp4psSkfUjyiiRXJ3kwydNJ9iW5NslJy9Hu\nOW2Z+P8wZz0XJakkX1yKdo7Y9lR9SLIxyWeSfLt/P301yTuXss0LtGOafeLEJJ/q30vPJHkgycUD\nH1+waJL8ZJLf7ffJp/v3xPqBdSffr6tqRU3AGrrvxd4H/Eu6uw9/me4rY2sH1L8WeIzutu3/FLiB\n7itppx0OfaD7nvBdwNuBLXS3j/8K8B3ghMOhD3PWswF4Evg/wBcPs/fSZuAJ4Ka+/tnAbwLvPlz6\nAayl+7bSHrpntJwNvK/fJ65b5n5s6d8HnwVupvuK6PqBdSfer5etg2P8Id4J/F/gJbPKTgIOjHpz\n0d1sr4A3zipbDXwVuOkw6cOL5ylbB/yQ7nvBK74Pc9ZzM3AlcMchCLlp/g9H0d1l54+Ws81L0I9/\n1u8T58wpv6Svv2YZ+3HUrJ/fPDTkpt2vV+LhagvPlJi4D1X1yDxl3wC+BfzEIrfzYKb5PwCQ5HXA\n6cBvL0kLR5umD1vo7lK74IOaltE0/XheP39sTvljdEGexWrkKFX1wwmrTrVfr8SQW1HPlJjQNH34\nEUk2Ai9meW9rPVUfkhwHfBh4X1UdqpujTtOHV/XzY5LcleQHSR5J8tEkL1jUVo42TT8+T3eoe2mS\nTUlemGQr3ejwiupuwLHSTbVfr8SQW6nPlBjHNH14jv7+e1fQjeSunr5pg03bh8vozgV9chHbNK5p\n+vDj/fw64Bbg1cCldIdZf7BYDRxo4n5U1ffoAvsoulD4LnAr8CfA2xa3mUtmqv36sHvGwxHoY8Ar\ngV+sqlG3m18Rkvws8OvA6dWfQDkMzQwAfr+qtvc/35FkFXBJko1VteIfGJPkGLqg/jG6B8PvA36a\n7mHvB+ju79i0lRhy0z5TYt0CdWH5nikxTR/+vySX0H2ad35V3bJIbRtqmj5cSTfq/Nskx/Zlq4FV\n/e/PVNWzi9bShU3Th+/08z+fU34L3Un701i+0wfT9OM36M4vnjzrnN5fJHkc+HiSK6pqpT/xbqr9\neiUerrbwTIlp+gBAkvcDvwW8o6quWcS2DTVNHzYCb6F7c85MZwFn9j8v1+hh2vfSSjFNP34KeGz2\nhxa9v+rnG1n5ptqvV2LI3QScmWTDTEF/weBZ/WsH88fA0cCvzKq7GngtcMsyjR5guj6Q5B10d1J+\nf1V9bInaOMo0fTh7nukeupPnZwPXL35z5zVNHz4HPEt36/7Zzu3nf704TRxkmn58Ezg2ydyT8z/T\nz/9ukdq4lKbbrw/1NUDzXBOzli6Zv0z38fg2uh3k68ALZy23ju6cwvY59T9NN1p4M91Fg9cD36M7\nP7Ti+wD8Gt01cZ+jG/nMnjYdDn1YYH13sPzXyU37Xvr3ffnvAD8PXER3AeonD5d+AOvpLmh+gL+/\nGPi9fdndzLp2bZn68sv99Ht01769tf/9n4z4X0y8Xy9b58b8Q5wI/GH/j/gu8D+Yc9Fg/88r4D/M\nKX8B3bVN3+z/CP8L2HK49IHu08haYLrjcOjDAuta9pBbhPdSgHf3AfN94BvADuDow6wfm4D/DjxE\nF9IP0H2z5rhD0I+DvreXYr/2GQ+SmrYSz8lJ0qIx5CQ1zZCT1DRDTlLTDDlJTTPkJDXNkJPUNENO\nUtMMOUlN+3+uNtSQiDFhfQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7ffaa0d471d0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gs = GridSpec(nrows=2, ncols=3, height_ratios=[20, 1])\n",
    "\n",
    "for cond in ['U', 'T', 'N']:\n",
    "    plt.figure(figsize=(15, 5))\n",
    "    \n",
    "    exp = expected[cond]\n",
    "    O = clrs[cond].matrix().fetch('chr19')\n",
    "    e = exp.loc['chr19']['balanced.avg']\n",
    "    E = toeplitz(e.values)\n",
    "    \n",
    "    ax = plt.subplot(gs[0])\n",
    "    img = ax.matshow(np.log10(X), cmap='fall', vmin=-4, vmax=-1.5)\n",
    "    cax = plt.subplot(gs[3])\n",
    "    plt.colorbar(img, cax=cax, orientation='horizontal')\n",
    "\n",
    "    ax = plt.subplot(gs[1])\n",
    "    img = ax.matshow(np.log10(E), cmap='fall', vmin=-4, vmax=-1.5)\n",
    "    cax = plt.subplot(gs[4])\n",
    "    plt.colorbar(img, cax=cax, orientation='horizontal')\n",
    "\n",
    "    ax = plt.subplot(gs[2])\n",
    "    img = ax.matshow(np.log10(O / E), cmap='coolwarm', vmin=-1, vmax=1)\n",
    "    cax = plt.subplot(gs[5])\n",
    "    plt.colorbar(img, cax=cax, orientation='horizontal')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "_draft": {
   "nbviewer_url": "https://gist.github.com/03cf7bc3440f7e41520910013ce9970d"
  },
  "gist": {
   "data": {
    "description": "projects/ismb-tutorial/02-expected.ipynb",
    "public": false
   },
   "id": "03cf7bc3440f7e41520910013ce9970d"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  },
  "toc": {
   "toc_cell": false,
   "toc_number_sections": true,
   "toc_threshold": 6,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
